File: clients/apple/Packages/PilotageCore/Sources/PilotageCore/HIDCharacterizationSampling.swift
import Foundation

/// A decoded Apple HID report for automatic device characterization.
public struct HIDCharacterizationSample: Codable, Equatable, Sendable {
    /// A capture-local sequence.
    public let sequence: UInt64
    /// Microseconds from capture start at bridge receipt.
    public let observedAtUs: UInt64
    /// Microseconds from the HID source clock, when available.
    public let sourceAtUs: UInt64?
    /// Axis values in raw HID units.
    public let axes: [Float]
    /// Raw report bytes in hexadecimal form, when available.
    public let reportHex: String?

    enum CodingKeys: String, CodingKey {
        case sequence
        case observedAtUs = "observed_at_us"
        case sourceAtUs = "source_at_us"
        case axes
        case reportHex = "report_hex"
    }
}

/// One guided capture segment.
public struct HIDCharacterizationSegment: Codable, Equatable, Sendable {
    /// The guided action.
    public let action: Action
    /// The first included sequence.
    public let startSequence: UInt64
    /// The last included sequence.
    public let endSequence: UInt64

    /// A guided operator action.
    public struct Action: Codable, Equatable, Sendable {
        /// `idle` or `movement`.
        public let kind: String
        /// The named control for a movement.
        public let logical: String?
        /// True when positive movement occurs first.
        public let positiveFirst: Bool?

        enum CodingKeys: String, CodingKey {
            case kind
            case logical
            case positiveFirst = "positive_first"
        }
    }

    enum CodingKeys: String, CodingKey {
        case action
        case startSequence = "start_sequence"
        case endSequence = "end_sequence"
    }
}

/// An Apple HID capture that uses the shared JSON schema.
public struct AppleHIDCharacterizationCapture: Codable, Equatable, Sendable {
    private static let maximumArtifactBytes = 64 * 1024 * 1024
    /// The schema version.
    public let schemaVersion: UInt32
    /// The USB device identity.
    public let device: Device
    /// A token for one source connection or one synthetic fixture instance.
    public let deviceInstanceID: String
    /// The sampling source name.
    public let source: String
    /// The selected timing clock.
    public let timestampSource: String
    /// The event represented by each timing sample.
    public let timingObservation: String
    /// Platform dead-zone evidence.
    public let deadzoneEvidence: DeadzoneEvidence
    /// SHA-256 of the exact source-axis contract.
    public let sourceContractDigest: String
    /// Trusted source-unit ranges for decoded axes.
    public let sourceAxes: [SourceAxis]
    /// Decoded reports.
    public let samples: [HIDCharacterizationSample]
    /// Guided segments.
    public let segments: [HIDCharacterizationSegment]

    /// A USB device identity.
    public struct Device: Codable, Equatable, Sendable {
        /// USB vendor ID.
        public let vendorID: UInt16
        /// USB product ID.
        public let productID: UInt16
        /// Product name, when available.
        public let product: String?

        enum CodingKeys: String, CodingKey {
            case vendorID = "vendor_id"
            case productID = "product_id"
            case product
        }

        /// Creates a USB device identity.
        public init(vendorID: UInt16, productID: UInt16, product: String?) {
            self.vendorID = vendorID
            self.productID = productID
            self.product = product
        }
    }

    /// Evidence about platform dead-zone shaping.
    public struct DeadzoneEvidence: Codable, Equatable, Sendable {
        /// The measured status.
        public let status: String
        /// The measurement method.
        public let method: String
        /// The report count.
        public let sampleCount: UInt64

        enum CodingKeys: String, CodingKey {
            case status
            case method
            case sampleCount = "sample_count"
        }
    }

    /// One trusted source-axis range and neutral position.
    public struct SourceAxis: Codable, Equatable, Sendable {
        /// The index in each decoded report.
        public let sourceIndex: Int
        /// The smallest source value.
        public let minimum: Float
        /// The largest source value.
        public let maximum: Float
        /// `centered`, `minimum`, or `maximum`.
        public let neutralPosition: String

        enum CodingKeys: String, CodingKey {
            case sourceIndex = "source_index"
            case minimum
            case maximum
            case neutralPosition = "neutral_position"
        }

        /// Creates one trusted source-axis contract entry.
        public init(sourceIndex: Int, minimum: Float, maximum: Float, neutralPosition: String) {
            self.sourceIndex = sourceIndex
            self.minimum = minimum
            self.maximum = maximum
            self.neutralPosition = neutralPosition
        }
    }

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case device
        case deviceInstanceID = "device_instance_id"
        case source
        case timestampSource = "timestamp_source"
        case timingObservation = "timing_observation"
        case deadzoneEvidence = "deadzone_evidence"
        case sourceContractDigest = "source_contract_digest"
        case sourceAxes = "source_axes"
        case samples
        case segments
    }

    /// Encodes one deterministic JSON artifact for digest and promotion input.
    public func encodedJSON() throws -> Data {
        try encodedJSON(maximumBytes: Self.maximumArtifactBytes)
    }

    func encodedJSON(maximumBytes: Int) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        var data = try encoder.encode(self)
        data.append(0x0A)
        guard data.count <= maximumBytes else { throw ArtifactError.encodedSizeLimit }
        return data
    }

    /// Errors from portable artifact encoding.
    public enum ArtifactError: Error, Equatable {
        /// The encoded capture exceeds the shared artifact limit.
        case encodedSizeLimit
    }
}

/// Collects caller-supplied samples for Apple schema interoperability tests.
///
/// The result uses synthetic provenance. Promotion rejects this provenance.
public struct AppleHIDCharacterizationFixtureSampler: Sendable {
    private static let maximumSamples = 1_000_000
    private static let maximumSegments = 65
    private static let maximumProductNameBytes = 256
    private static let maximumLogicalNameBytes = 64
    private let device: AppleHIDCharacterizationCapture.Device
    private let deviceInstanceID: String
    private let sourceContractDigest: String
    private let sourceAxes: [AppleHIDCharacterizationCapture.SourceAxis]
    private var samples: [HIDCharacterizationSample] = []
    private var segments: [HIDCharacterizationSegment] = []
    private var openAction: HIDCharacterizationSegment.Action?
    private var openStart: UInt64 = 0

    /// Creates a sampler bound to one synthetic device instance and source contract.
    public init(
        device: AppleHIDCharacterizationCapture.Device,
        deviceInstanceID: String,
        sourceContractDigest: String,
        sourceAxes: [AppleHIDCharacterizationCapture.SourceAxis]
    ) throws {
        guard device.product.map({ !$0.isEmpty && $0.utf8.count <= Self.maximumProductNameBytes })
            ?? true
        else { throw SamplingError.invalidProductName }
        guard !deviceInstanceID.isEmpty, deviceInstanceID.utf8.count <= 256 else {
            throw SamplingError.invalidDeviceInstance
        }
        guard Self.isDigest(sourceContractDigest) else {
            throw SamplingError.invalidSourceContract
        }
        guard !sourceAxes.isEmpty, sourceAxes.count <= 64 else {
            throw SamplingError.invalidSourceContract
        }
        for (index, axis) in sourceAxes.enumerated() {
            guard axis.sourceIndex == index, axis.minimum.isFinite, axis.maximum.isFinite,
                  axis.minimum < axis.maximum,
                  ["centered", "minimum", "maximum"].contains(axis.neutralPosition)
            else { throw SamplingError.invalidSourceContract }
        }
        self.device = device
        self.deviceInstanceID = deviceInstanceID
        self.sourceContractDigest = sourceContractDigest
        self.sourceAxes = sourceAxes
    }

    /// Starts the idle segment.
    public mutating func beginIdle() throws {
        try begin(.init(kind: "idle", logical: nil, positiveFirst: nil))
    }

    /// Starts one named positive-first movement segment.
    public mutating func beginMovement(logical: String) throws {
        guard !logical.isEmpty, logical.utf8.count <= Self.maximumLogicalNameBytes else {
            throw SamplingError.invalidLogicalName
        }
        try begin(.init(kind: "movement", logical: logical, positiveFirst: true))
    }

    /// Records one caller-supplied sample.
    public mutating func record(
        deviceInstanceID: String,
        isConnected: Bool,
        axes: [Float],
        observedAtUs: UInt64,
        sourceAtUs: UInt64?,
        reportHex: String?
    ) throws {
        guard openAction != nil else { throw SamplingError.noOpenSegment }
        guard isConnected else { throw SamplingError.deviceDisconnected }
        guard deviceInstanceID == self.deviceInstanceID else {
            throw SamplingError.deviceInstanceChanged
        }
        guard samples.count < Self.maximumSamples else { throw SamplingError.sampleLimit }
        guard reportHex == nil else {
            throw SamplingError.invalidReport
        }
        guard axes.count == sourceAxes.count,
              axes.enumerated().allSatisfy({ index, value in
                  value.isFinite && value >= sourceAxes[index].minimum
                      && value <= sourceAxes[index].maximum
              })
        else {
            throw SamplingError.invalidAxes
        }
        guard samples.last.map({ observedAtUs > $0.observedAtUs }) ?? true else {
            throw SamplingError.nonMonotonicTimestamp
        }
        if let sourceAtUs, let previous = samples.last?.sourceAtUs, sourceAtUs <= previous {
            throw SamplingError.nonMonotonicTimestamp
        }
        samples.append(
            .init(
                sequence: UInt64(samples.count),
                observedAtUs: observedAtUs,
                sourceAtUs: sourceAtUs,
                axes: axes,
                reportHex: reportHex
            )
        )
    }

    /// Closes the current segment.
    public mutating func endSegment() throws {
        guard let action = openAction else { throw SamplingError.noOpenSegment }
        guard UInt64(samples.count) > openStart else { throw SamplingError.emptySegment }
        guard segments.count < Self.maximumSegments else { throw SamplingError.segmentLimit }
        segments.append(
            .init(action: action, startSequence: openStart, endSequence: UInt64(samples.count - 1))
        )
        openAction = nil
    }

    /// Creates a portable capture for the bound device.
    public func finish() throws -> AppleHIDCharacterizationCapture {
        guard openAction == nil else { throw SamplingError.segmentStillOpen }
        guard !samples.isEmpty, !segments.isEmpty else { throw SamplingError.emptyCapture }
        let sourceClock = samples.allSatisfy { $0.sourceAtUs != nil }
        let capture = AppleHIDCharacterizationCapture(
            schemaVersion: 1,
            device: device,
            deviceInstanceID: deviceInstanceID,
            source: "synthetic",
            timestampSource: sourceClock ? "source" : "arrival",
            timingObservation: "injected_samples",
            deadzoneEvidence: .init(
                status: "unknown",
                method: "unmeasured",
                sampleCount: 0
            ),
            sourceContractDigest: sourceContractDigest,
            sourceAxes: sourceAxes,
            samples: samples,
            segments: segments
        )
        _ = try capture.encodedJSON()
        return capture
    }

    private mutating func begin(_ action: HIDCharacterizationSegment.Action) throws {
        guard openAction == nil else { throw SamplingError.segmentStillOpen }
        openAction = action
        openStart = UInt64(samples.count)
    }

    private static func isDigest(_ value: String) -> Bool {
        value.utf8.count == 64 && value.utf8.allSatisfy { byte in
            (48 ... 57).contains(byte) || (97 ... 102).contains(byte)
        }
    }

    /// Errors from capture sequencing and sample validation.
    public enum SamplingError: Error, Equatable {
        /// A segment is already open.
        case segmentStillOpen
        /// No segment is open.
        case noOpenSegment
        /// A segment has no report.
        case emptySegment
        /// A movement name is empty or too large.
        case invalidLogicalName
        /// The synthetic sampler received a raw report.
        case invalidReport
        /// The device product name is empty or too large.
        case invalidProductName
        /// Axis values are empty or non-finite.
        case invalidAxes
        /// The caller declared that the synthetic device is disconnected.
        case deviceDisconnected
        /// A sample came from a different synthetic device instance.
        case deviceInstanceChanged
        /// The device instance token is invalid.
        case invalidDeviceInstance
        /// The source-axis contract is invalid.
        case invalidSourceContract
        /// The capture reached its report limit.
        case sampleLimit
        /// The capture reached its segment limit.
        case segmentLimit
        /// Arrival timestamps do not increase.
        case nonMonotonicTimestamp
        /// The capture has no report or segment.
        case emptyCapture
    }
}

File: clients/web/hid-characterization-sampling.js
// Browser Gamepad sampling for the shared HID characterization schema.
import { assertGamepadBinding, bindGamepadContract, isWellFormedText } from "./hid-characterization-contract.js";
const SCHEMA_VERSION = 1;
const MAX_SAMPLES = 1_000_000;
const MAX_SEGMENTS = 65;
const MAX_LOGICAL_NAME_BYTES = 64;
const MAX_CAPTURE_BYTES = 64 * 1024 * 1024;
/** Records changed browser Gamepad states without claiming raw HID loss. */
export class BrowserHIDCharacterizationSampler {
  constructor(options) {
    this.binding = bindGamepadContract(options);
    this.samples = [];
    this.segments = [];
    this.openSegment = null;
    this.lastSourceTimestampMs = null;
    this.lastObservedAtUs = null;
  }
  /** Starts the one required idle segment. */
  beginIdle() {
    this.#begin({ kind: "idle" });
  }

  /** Starts one named movement. The positive direction must move first. */
  beginMovement(logical) {
    if (!isWellFormedText(logical) || logical.length === 0 ||
        new TextEncoder().encode(logical).length > MAX_LOGICAL_NAME_BYTES) {
      throw new TypeError("movement logical name is invalid");
    }
    this.#begin({ kind: "movement", logical, positive_first: true });
  }

  /** Records one changed Gamepad state. Returns false for a repeated timestamp. */
  record(gamepad, connectionEpoch, observedAtUs) {
    if (this.openSegment === null) throw new Error("start a segment before sampling");
    assertGamepadBinding(this.binding, gamepad, connectionEpoch);
    if (!Number.isSafeInteger(observedAtUs) || observedAtUs < 0) {
      throw new TypeError("observedAtUs must be a non-negative integer");
    }
    if (this.lastObservedAtUs !== null && observedAtUs <= this.lastObservedAtUs) {
      throw new RangeError("arrival timestamps must increase");
    }
    const sourceMs = Number(gamepad.timestamp);
    const sourceUs = Math.round(sourceMs * 1000);
    if (!Number.isFinite(sourceMs) || sourceMs < 0 || !Number.isSafeInteger(sourceUs)) {
      throw new TypeError("Gamepad.timestamp is required for state-update timing");
    }
    if (sourceMs === this.lastSourceTimestampMs) return false;
    if (this.lastSourceTimestampMs !== null && sourceMs < this.lastSourceTimestampMs) {
      throw new RangeError("Gamepad timestamps must not decrease");
    }
    if (this.samples.length >= MAX_SAMPLES) throw new RangeError("capture sample limit reached");
    const axes = Array.from(gamepad.axes, Number);
    if (axes.some((axis) => !Number.isFinite(axis) || axis < -1 || axis > 1)) {
      throw new TypeError("Gamepad axes must be finite values in [-1, 1]");
    }
    this.samples.push({
      sequence: this.samples.length,
      observed_at_us: observedAtUs,
      source_at_us: sourceUs,
      axes,
      report_hex: null,
    });
    this.lastObservedAtUs = observedAtUs;
    this.lastSourceTimestampMs = sourceMs;
    return true;
  }

  /** Closes the current segment after at least one changed state. */
  endSegment() {
    if (this.openSegment === null) throw new Error("no segment is open");
    if (this.samples.length === this.openSegment.sampleStart) {
      throw new Error("a segment needs at least one changed state");
    }
    if (this.segments.length >= MAX_SEGMENTS) throw new RangeError("capture segment limit reached");
    this.segments.push({
      action: this.openSegment.action,
      start_sequence: this.openSegment.sampleStart,
      end_sequence: this.samples.length - 1,
    });
    this.openSegment = null;
  }

  /** Returns a JSON-ready portable capture. */
  finish(maximumCaptureBytes = MAX_CAPTURE_BYTES) {
    if (!Number.isSafeInteger(maximumCaptureBytes) || maximumCaptureBytes <= 0 || maximumCaptureBytes > MAX_CAPTURE_BYTES) throw new RangeError("invalid encoded capture byte limit");
    if (this.openSegment !== null) throw new Error("close the current segment first");
    if (this.samples.length === 0 || this.segments.length === 0) {
      throw new Error("the capture is empty");
    }
    const binding = this.binding;
    const capture = {
      schema_version: SCHEMA_VERSION,
      device: { ...binding.device },
      device_instance_id: binding.connectionEpoch,
      source: "browser_gamepad",
      timestamp_source: "source",
      timing_observation: "polled_state_updates",
      deadzone_evidence: { status: "unknown", method: "unmeasured", sample_count: 0 },
      source_contract_digest: binding.sourceContractDigest,
      source_axes: binding.sourceAxes.map((axis) => ({ ...axis })),
      samples: this.samples.map((sample) => ({ ...sample, axes: [...sample.axes] })),
      segments: this.segments.map((segment) => ({
        ...segment,
        action: { ...segment.action },
      })),
    };
    if (new TextEncoder().encode(`${JSON.stringify(capture)}\n`).byteLength > maximumCaptureBytes) throw new RangeError("encoded capture byte limit reached");
    return capture;
  }

  #begin(action) {
    if (this.openSegment !== null) throw new Error("close the current segment first");
    this.openSegment = { action, sampleStart: this.samples.length };
  }
}
